src/iceberg/type.h

```diff
@@ -0,0 +1,397 @@
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 32-bit (single precision) float.
+class ICEBERG_EXPORT Float32Type : public PrimitiveType {
+ public:
+  Float32Type() = default;
+  ~Float32Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a 64-bit (double precision) float.
+class ICEBERG_EXPORT Float64Type : public PrimitiveType {
+ public:
+  Float64Type() = default;
+  ~Float64Type() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a fixed-precision decimal.
+class ICEBERG_EXPORT DecimalType : public PrimitiveType {
+ public:
+  constexpr static const int32_t kMaxPrecision = 38;
+
+  /// \brief Construct a decimal type with the given precision and scale.
+  DecimalType(int32_t precision, int32_t scale);
+  ~DecimalType() = default;
+
+  /// \brief Get the precision (the number of decimal digits).
+  [[nodiscard]] int32_t precision() const;
+  /// \brief Get the scale (essentially, the number of decimal digits after
+  ///   the decimal point; precisely, the value is scaled by $$10^{-s}$$.).
+  [[nodiscard]] int32_t scale() const;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+
+ private:
+  int32_t precision_;
+  int32_t scale_;
+};
+
+/// \brief A data type representing a calendar date without reference to a
+///   timezone or time.
+class ICEBERG_EXPORT DateType : public PrimitiveType {
+ public:
+  DateType() = default;
+  ~DateType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a wall clock time in microseconds without
+///   reference to a timezone or date.
+class ICEBERG_EXPORT TimeType : public PrimitiveType {
+ public:
+  TimeType() = default;
+  ~TimeType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a timestamp in microseconds without
+///   reference to a timezone.
+class ICEBERG_EXPORT TimestampType : public PrimitiveType {
+ public:
+  TimestampType() = default;
+  ~TimestampType() = default;
+
+  TypeId type_id() const override;
+  std::string ToString() const override;
+
+ protected:
+  bool Equals(const Type& other) const override;
+};
+
+/// \brief A data type representing a timestamp in microseconds in a
+///   particular timezone.
+class ICEBERG_EXPORT TimestampTzType : public PrimitiveType {
```

Comment:
Should we add an abstract base class for all timestamp types? We will need `timestamp_ns` and `timestamptz_ns` for v3. The only differences are `timezone` and `precision`.

Reply:
Are timezone types parameterized? The Java implementation is weird and only has `adjustToUTC` (so I guess there are only naive and UTC timestamps in Iceberg?): https://github.com/apache/iceberg/blob/ab9e05e7369c55a95632b856729a05ea876cf788/api/src/main/java/org/apache/iceberg/types/Types.java#L211

Reply:
Right, Parquet and Iceberg only care about whether this is normalized to UTC.

Reply:
Hmm, I can add a base class then.